### ISSUE TITLE
Bridge.net support for thread C#

### ISSUE THREAD

**Issue by pavan1009** [bridgedotnet/Bridge#3923]
Desk Link : https://deck.net/f3a132eb6c60d59e1f96c058651a9310

trying to convert c# to JS using bridge.net but failed as system.threading is not fully supported in Bridge as in C#.
Please have a look on deck link.

**Comment by ChrML** [comment]
This would be nice, but unfortunately true multithreading is extremely difficult to accomplish using ECMA/Javascript.

The newest webstandards does support multiple threads, but only through WebWorker's. These workers run in true OS threads, but to avoid having to deal with race-conditions, deadlocks and other concurrency issues normal to multithreaded applications, it was decided that the webworkers shall run in a totally seperare context with no access to shared variables or the DOM. The only way to communicate to/from workers is by posting events. See more info about them here:
https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers

In C#/.NET it is assumed that all threads have direct access to all data, and that it's up to the programmer to prevent concurrency issues. Hence the System.Threading.Thread and WebWorker philosophies are directly incompatible by concept. 

I'm not sure if threading actually could be faked either. I guess it could be possible by rewriting Bridge. NET to maintain task scheduling, and then implement some sort of automatic generated cooperative threading system. This would be interesting, but I'm not sure the results would be any good, and it could make apps unnecessary complex for no good reason except direct portability (with sideeffect). In fact rewriting/porting the applications to not use threads would be much better.

You should note that the .NET Task- pattern (async/await) operates in a similar singlethreaded mode like that. Putting your operations into tasks may ease the porting.

An interesting fact is that the WebWorker approach discussed above is actually quite similar to the AppDomain concept in .NET that has the exact same concept of no datasharing. Data has to be copied between appdomains using serialization, or marshalled using MarshalByRef. This exact same behaviour could be implemented in a future version of Bridge.NET using WebWorkers. In some cases where you need to do CPU intensive tasks related to communication with the webserver without affecting the UI that would be great, then just serialize the result and post it to the main javascript thread when it's completed to update the webpage.

**Comment by miladsadeghi72** [comment]
Web workers is good but in worker we can't access direct to page components
Also we can't access to bridge library, so the threads in bridge is imposible
So we can just write Javascript for webworker

Otherwise i think we can make thread in js standalone, it's possible via make code tracer, but I don't know about performance 

So you right, now async task is best way and so optimum in js and bridge c# also

**Comment by ChrML** [comment]
@miladsadeghi72 Actually you can use Bridge.NET inside webworkers, altough you need to load the Bridge js again inside the webworker. But like you say you cannot access the DOM or the scope from the main thread. Hence you're not limited to pure js.

It's the very same concept as AppDomain in .NET. If you want to post data from the worker to the main thread or vice versa you need to serialize the data (NewtonsoftJSON or your own serializer). And you need to implement MarshalByRef to communicate with reference types across the app domains (which are actually webworkers).

But webworkers will never help implementing Bridge Thread identical to .NET threads, so kinda offtopic.

**Comment by athendrix** [comment]
Just because we can't do multi-threading the exact same way as standard .NET doesn't mean we shouldn't add some sort of Web Worker support. Perhaps we could implement the support with a class that can be inherited and instantiated to create a new worker with Bridge.NET. We could have override methods for onmessage or whatever, and just let any attempts to cross the Web Worker border throw some kind of Web Worker Exceptions.

Bridge.NET could create the Web Worker using a blob url so it can change details at runtime.

How feasible does this sound do you think? Am I off my rocker?

**Comment by ChrML** [comment]
@athendrix You are right, this is possible and not very hard to do. I considered doing it myself to speed up the UI when handling larger XML data, but postponed it.

One could use the WebWorker class to implement the AppDomain model later. Those two models are very fundamentally compatible, by having the same restrictions.
